finrl/env/env_stocktrading_cashpenalty.py

```diff
@@ -288,10 +288,17 @@ def step(self, actions):
             costs += spend * self.buy_cost_pct
              # if we run out of cash...
             if (spend + costs) > coh:
-                # ... end the cycle and penalize
-                return self.return_terminal(
-                    reason="CASH SHORTAGE",reward=self.get_reward()
-                )
+                if self.patient:
+                    # ... just don't buy anything until we got additional cash
+                    self.log_step(reason="CASH SHORTAGE")
+                    actions = np.where(actions>0,0,actions)
+                    spend = 0
+                    costs = 0
+                else:
+                    # ... end the cycle and penalize
+                    return self.return_terminal(
+                        reason="CASH SHORTAGE",reward=self.get_reward()
+                    )
             self.transaction_memory.append(actions)  # capture what the model's could do
```

Comment:
This should be before any modifications to action, so please put this higher in the step function. Actions == raw decisions from model. Transactions=  actual actions applied in this step.

Reply:
I think that's transaction_memory not action_memory, so it should be okay to put it down there to also capture the changes of actions if the user turns on the patient parameter. What do you think?

Reply:
oh, yes, sorry you're right!